### PR TITLE
api: add cluster_specifier_plugin to RouteAction

### PR DESCRIPTION
**Pull request by dfawley**
Commit Message: api: add cluster_specifier_plugin to RouteAction
Additional Description:

This is a new extension point that will allow for dynamic cluster selection.
The plugin configured by it will be responsible for returning a cluster to
Envoy through its API (to be defined).  Since the cluster returned is dynamic,
a new field is added to Route to list all the clusters it may return.  This
allows for proxy implementations that are not using wildcard CDS queries to
pre-fetch clusters.

This feature may ultimately replace the FilterAction mechanism, as it provides
the same functionality for known use cases, but is simpler to implement and
use.

Risk Level: Low
Testing: N/A
Docs Changes: Included in PR.
Release Notes: N/A
Platform Specific Features: N/A

cc @htuch, @markdroth

Related to #8953, #8956

### CODE REVIEW

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:307]: I think this field should be in `RouteConfiguration`, not in `Route`.  That way, if there are many routes that use the same `cluster_specifier_plugin`, we don't need to specify this list of additional clusters multiple times.

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:306]: Might be a good idea to explicitly state that this is intended for clients (like gRPC) that fetch the specific set of cluster resources referred to in the RouteConfiguration rather than making a wildcard CDS subscription.

**Review by markdroth — CHANGES_REQUESTED** [review]
Thanks for putting this together!

**Comment by markdroth** [comment]
This looks great to me!

@mattklein123, can you provide a non-Google review here?  For context, the reason why we're coming back to this approach instead of using `FilterAction` is that, while `FilterAction` is a more general solution, it also turns out to be much more complicated to implement.  We have several use-cases right now that need to dynamically determine the cluster, but we don't have any use-cases that I'm aware of that actually need the additional generality of `FilterAction`.  Given all of that, we'd prefer to go with this approach.

Please let us know if you have any questions.  Thanks!

File: api/envoy/config/route/v3/route.proto
Comment (htuch) [inline api/envoy/config/route/v3/route.proto:125]: Nit: add a RST link to `cluster_specifier`, `*RouteAction*` etc.

**Review by htuch — COMMENTED** [review]
FWIW I'm good with this PR, but @mattklein123 should provide non-Google coverage as next step.

File: api/envoy/config/route/v3/route.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route.proto:129]: One thought: WDYT about colocating this with `cluster_specifier_plugin` perhaps in a wrapper message? When the client loads the route it could just as easily keep a set of additional_clusters that it comes across? Another option would be to elide this config entirely and just have the additional clusters concept be part of the plugin API that is read during load, though I understand that is probably not desirable given the desire to separate code from config.

**Review by mattklein123 — COMMENTED** [review]
In general this makes sense to me. One thought.

/wait-any

File: api/envoy/config/route/v3/route.proto
Comment (dfawley) [reply api/envoy/config/route/v3/route.proto:129]: I'm definitely open to adding a wrapper message, and optionally including this field in it.  (Incidentally, I like the idea of a wrapper message anyway for a possible extension point, but I'll defer to whatever the convention is for Envoy.)  However, the reason this is in `RouteConfiguration` is to avoid the situation where many different routes all reference the same plugin, which is capable of returning a large set of clusters (I've found it's ~1.5kb in practice for one data point).  This could lead to config bloat, so we decided to move it here to be able to deduplicate.

Comment (mattklein123) [reply api/envoy/config/route/v3/route.proto:129]: > This could lead to config bloat, so we decided to move it here to be able to deduplicate.

Fair enough, but then do you need the cluster_specifier_plugin on each route? Could it all just live at the route level and we can add vhost/route overrides later if needed?

Comment (markdroth) [reply api/envoy/config/route/v3/route.proto:129]: I don't think that putting the actual plugin config at the top level would work, because you might need different plugins for different routes.

If the plugin config itself gets too large and we don't want to repeat it in multiple routes, we can move the bulk of the config to an ECDS resource, so that the only thing we need to include in each route is a `ConfigSource` pointing to the ECDS resource.

Comment (mattklein123) [reply api/envoy/config/route/v3/route.proto:129]: I think my point is that saying that one part of this might get too big for the route while the other one might not sounds strange to me. Why not just have a wrapper message and put it at the top level, and then also allow overriding it at the vhost or route level? This is how we do most other things.

Comment (dfawley) [reply api/envoy/config/route/v3/route.proto:129]: > I think my point is that saying that one part of this might get too big for the route while the other one might not sounds strange to me.

One key difference is that the config of the plugin is opaque to xDS clients, but the additional clusters needs to be understood.

Another option would be:

- In `RouteConfiguration`, a map from string (specifier plugin instance name) to `ClusterSpecifierPluginConfig` which includes a field for `additional_clusters`.
- Make `cluster_specifier_plugin` a string that references that map.

Comment (mattklein123) [reply api/envoy/config/route/v3/route.proto:129]: Yeah I think ^ would work. I would definitely do the wrapper message. What I'm saying though is to do the following:
1) `ClusterSpecifierPluginConfig` at route config level.
2) `ClusterSpecifierPluginConfig` at individual route level (and maybe vhost level)
3) Most specific `ClusterSpecifierPluginConfig` wins. This is probably generally wire efficient enough for almost all uses cases without the indirection.

Comment (markdroth) [reply api/envoy/config/route/v3/route.proto:129]: Matt, I think your point about not treating these two pieces of data separately is well taken.  As it happens, we just realized that we don't actually need the list of clusters anyway, so we're now back to just having the plugin config.  However, I agree that it would be nice to have a way to avoid duplicating the plugin config in each route without having to use ECDS.  So I think you're right that we should find a way to move that to the top level of the `RouteConfiguration`.

That having been said, I am thinking that we should design this in a way that makes it efficient to have multiple plugins (we don't need this now but might in the future), each of which is used in multiple routes.  For example, consider a case where we have two plugins defined, `plugin1` and `plugin2`, and we have the following list of routes:

1. cluster=A
2. cluster_specifier_plugin=plugin1
3. cluster_specifier_plugin=plugin2
4. cluster_specifier_plugin=plugin1
5. cluster_specifier_plugin=plugin2
6. cluster_specifier_plugin=plugin1
7. cluster_specifier_plugin=plugin2

If we do this as a single `ClusterSpecifierPluginConfig` per `RouteConfiguration` that can be overridden in individual routes, then we would need to set the top-level `ClusterSpecifierPluginConfig` to refer to `plugin1`, and then we'd need to duplicate the `ClusterSpecifierPluginConfig` for `plugin2` in the three routes that need to use `plugin2` instead of `plugin1`.  Also, even for the routes that do use `plugin1`, we'd still need a new field in the `cluster_specifier` oneof in the individual routes to indicate that it should use the top-level `ClusterPerSpecifierPluginConfig` instead of specifying a particular cluster.  So we'd wind up adding two new fields in each route, like so:

```
oneof cluster_specifier {
  // Existing fields.
  string cluster = N;
  string cluster_header = N;
  WeightedClusters weighted_clusters = N;

  // New field to indicate using the top-level ClusterSpecifierPluginConfig.
  bool use_plugin = N;
  // New field to allow overriding the ClusterSpecifierPluginConfig.
  ClusterSpecifierPluginConfig plugin_config = N;
}
```

And the resulting `RouteConfiguration` would look like this, with the `plugin2` config duplicated three times (textproto format):

```
// Top-level field in RouteConfig
cluster_specifier_plugin: { name: "plugin1" typed_config: { [envoy.extensions.cluster_specifier.plugin1.Plugin1] { ... } } }

// ...

// Route 1
cluster: "A"
// Route 2
use_plugin: true
// Route 3
plugin_config: { name: "plugin2" typed_config: { [envoy.extensions.cluster_specifier.plugin2.Plugin2] { ... } } }
// Route 4
use_plugin: true
// Route 5
plugin_config: { name: "plugin2" typed_config: { [envoy.extensions.cluster_specifier.plugin2.Plugin2] { ... } } }
// Route 6
use_plugin: true
// Route 7
plugin_config: { name: "plugin2" typed_config: { [envoy.extensions.cluster_specifier.plugin2.Plugin2] { ... } } }
```

This duplicates the `plugin2` config 3 times, which is both inefficient and somewhat counter-intuitive, especially since there will generally be only one actual instance of `plugin2`.

I think we can avoid a lot of that complexity by specifying the list of plugins at the top-level, keyed by instance name, and then referring to those plugins by name in the individual routes.  The idea would be to have something like this in the top-level of the `RouteConfiguration`:

```
// A list of cluster_specifier plugin instances.  All names in the list must be unique.
// Note: Ideally, this would be a map, but the map key would just duplicate the name field in TypedExtensionConfig.
repeated core.v3.TypedExtensionConfig cluster_specifier_plugin_instances = N;
```

And then all we need in the `cluster_specifier` oneof is a single new field:

```
// Refers to the plugins defined in the RouteConfiguration top-level cluster_specifier_plugin_instances field.
string cluster_specifier_plugin_name = N;
```

That way, we can configure multiple plugins like this (textproto format):

```
cluster_specifier_plugin_instances: [
  { name: "plugin1", typed_config: { [envoy.extensions.cluster_specifier.plugin1.Plugin1] { ... } } },
  { name: "plugin2", typed_config: { [envoy.extensions.cluster_specifier.plugin2.Plugin2] { ... } } }
]
```

And then individual routes could specify either "plugin1" or "plugin2", and there could be multiple routes referring to each of the two plugins.

WDYT?

Comment (mattklein123) [reply api/envoy/config/route/v3/route.proto:129]: Yup ^ sounds great. Let's do it!

Comment (dfawley) [reply api/envoy/config/route/v3/route.proto]: Done with minor style changes, and included a wrapper message for future proofing - PTAL.

File: api/envoy/config/route/v3/route.proto
Comment (markdroth) [inline api/envoy/config/route/v3/route.proto:126]: I think this should be:

```
:ref:`cluster_specifier_plugin <envoy_v3_api_field_config.route.v3.RouteAction.cluster_specifier_plugin>`
```

**Review by markdroth — COMMENTED** [review]
This looks great!  Just one minor nit in the doc comment to address.

File: api/envoy/config/route/v3/route.proto
Comment (markdroth) [reply api/envoy/config/route/v3/route.proto:126]: Looks like this actually works fine the way you have it, so nevermind. :)

**Comment by markdroth** [comment]
I'll let @mattklein123 do the approval here, but this looks great!

File: api/envoy/config/route/v3/route.proto
Comment (dfawley) [reply api/envoy/config/route/v3/route.proto:126]: Discussed offline: we think this should be fine as-is, actually.  I can change it easily if needed, though.  (I already made the edits locally.)

**Comment by alyssawilk** [comment]
hey @markdroth Matt is on leave this week, so would you be up for approving or throwing over to another shephard if you wanted a second pair of eyes?

**Comment by markdroth** [comment]
The impetus for this comes from here at Google, so I'd prefer to have a non-Google API shepherd give the final approval.

@lizan, can you please take a look?  Thanks!

File: api/envoy/config/route/v3/route.proto
Comment (lizan) [inline api/envoy/config/route/v3/route.proto:128]: shall we use `map<string, ClusterSpecifierPlugin>` here? That guarantees uniqueness of the key, and `extension.name` is historically used for specifying kind of extension, which is now implied by type, so that avoids confusion.

Comment (markdroth) [reply api/envoy/config/route/v3/route.proto:128]: I believe that nowadays, `extension.name` is intended to be used to specify the instance name, not the type of the extension.  This is reflected in the comment on that field:

https://github.com/envoyproxy/envoy/blob/c1bc5e78fa68b86236c8d6237e2db15ce1743459/api/envoy/config/core/v3/extension.proto#L22

This allows (for example) having multiple instances of the same type of extension under different names and then deciding which instance to use where.

Given that, I think using a map here would bloat the config unnecessarily by duplicating these instance names.

An alternative would be to change `ClusterSpecifierPlugin` to directly contain an `Any` field instead of using `TypedExtensionConfig`, so that we could put the name in the map here and not duplicate it.  But I think that we're trying to move toward using `TypedExtensionConfig` everywhere instead of bare `Any` fields, aren't we?

Comment (lizan) [reply api/envoy/config/route/v3/route.proto:128]: Yes nowadays it is used to specify instance name.

I'm ok either way but we would need to do duplicate key check by using repeated.

Comment (markdroth) [reply api/envoy/config/route/v3/route.proto:128]: Yeah, I think doing the duplicate key check is fine.  The client implementation will need to put each key into a map anyway, so we can just check to make sure that the key is not already present right before we add it.

Comment (htuch) [reply api/envoy/config/route/v3/route.proto:128]: Just a +1 to where we ended up (@markdroth asked for my opinion offline). I don't think it matters greatly, but DRY seems best (once again the weakness of protobuf type system strikes).

File: api/envoy/config/route/v3/route_components.proto
Comment (mattklein123) [inline api/envoy/config/route/v3/route_components.proto:846]: nit: pedantically, can you mention that this string refers to the name inside the `TypedExtensionConfig`? It's not completely clear to the reader what this string refers to.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small nit.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

File: api/envoy/config/route/v3/route_components.proto
Comment (dfawley) [reply api/envoy/config/route/v3/route_components.proto]: FWIW I'm not 100% sure about the refs -- it _looks_ like `envoy_v3_api_msg_config` is used to link to a message and `envoy_v3_api_field_config` is used to link to a field within a message.  Is that right?

Also I'm having some trouble running `fix_format`, so this should not be merged until I've worked that out.  (I assume the presubmits will fail anyway.)

**Review by mattklein123 — APPROVED** [review]

**Comment by dfawley** [comment]
> envoy-presubmit (check linux_x64 coverage) Failing after 72m — check linux_x64 coverage failed

This doesn't seem legit -- is it? If not, is there a way to re-run it without invalidating the approval?

**Comment by mattklein123** [comment]
> This doesn't seem legit -- is it? If not, is there a way to re-run it without invalidating the approval?

Seems like a flake. I will force merge.

**Pull request merged**
